libraries/AP_GPS/AP_GPS.cpp

```diff
@@ -883,19 +1050,440 @@ void AP_GPS::inject_data_all(const uint8_t *data, uint16_t len)
+    if (instance == GPS_MAX_RECEIVERS) {
+        return _blended_lag_sec;
+    }
+
     if (_delay_ms[instance] > 0) {
         // if the user has specified a non zero time delay, always return that value
         return 0.001f * (float)_delay_ms[instance];
     } else if (drivers[instance] == nullptr || state[instance].status == NO_GPS) {
         // no GPS was detected in this instance
         // so return a default delay of 1 measurement interval
-        return 0.001f * (float)_rate_ms[instance];
+        return 0.001f * (float)get_rate_ms(instance);
     } else {
         // the user has not specified a delay so we determine it from the GPS type
         return drivers[instance]->get_lag();
     }
 }
+
+// return a 3D vector defining the offset of the GPS antenna in meters relative to the body frame origin
+const Vector3f &AP_GPS::get_antenna_offset(uint8_t instance) const
+{
+    if (instance == GPS_MAX_RECEIVERS) {
+        // return an offset for the blended GPS solution
+        return _blended_antenna_offset;
+    } else {
+        return _antenna_offset[instance];
+    }
+}
+
+/*
+  return gps update rate in milliseconds
+*/
+uint16_t AP_GPS::get_rate_ms(uint8_t instance) const
+{
+    // sanity check
+    if (instance >= num_instances || _rate_ms[instance] <= 0) {
+        return GPS_MAX_RATE_MS;
+    }
+    return MIN(_rate_ms[instance], GPS_MAX_RATE_MS);
+}
+
+/*
+ calculate the weightings used to blend GPs location and velocity data
+*/
+bool AP_GPS::calc_blend_weights(void)
+{
+    // zero the blend weights
+    memset(&_blend_weights, 0, sizeof(_blend_weights));
+
+    // exit immediately if not enough receivers to do blending
+    if (num_instances < 2) {
+        return false;
+    }
+
+    // Use the oldest non-zero time, but if time difference is excessive, use newest to prevent a disconnected receiver from blocking updates
+    uint32_t max_ms = 0; // newest non-zero system time of arrival of a GPS message
+    uint32_t min_ms = -1; // oldest non-zero system time of arrival of a GPS message
+    int16_t max_rate_ms = 0; // largest update interval of a GPS receiver
+    for (uint8_t i=0; i<GPS_MAX_RECEIVERS; i++) {
+        // Find largest and smallest times
+        if (state[i].last_gps_time_ms > max_ms) {
+            max_ms = state[i].last_gps_time_ms;
+        }
+        if ((state[i].last_gps_time_ms < min_ms) && (state[i].last_gps_time_ms > 0)) {
+            min_ms = state[i].last_gps_time_ms;
+        }
+        if (get_rate_ms(i) > max_rate_ms) {
+            max_rate_ms = get_rate_ms(i);
+        }
+    }
+    if ((int32_t)(max_ms - min_ms) < (int32_t)(2 * max_rate_ms)) {
+        // data is not too delayed so use the oldest time_stamp to give a chance for data from that receiver to be updated
+        state[GPS_MAX_RECEIVERS].last_gps_time_ms = min_ms;
+    } else {
+        // receiver data has timed out so fail out of blending
+        state[GPS_MAX_RECEIVERS].last_gps_time_ms = max_ms;
```

Comment:
It doesn't matter. We are going to fall into the alternate switch to best GPS logic, at the end of which we will copy the current instance over the top of the state instance, so we will be replacing this anyways, this is basically just a no-op.